benches/mod.rs

```diff
@@ -1,94 +1,68 @@
 #![feature(test)]
 extern crate test;
 
-use std::{
-    alloc::{alloc_zeroed, dealloc, Layout},
-    ptr::NonNull,
-};
-
-// AlignedBuffer is like a Box<[u8; N]> except that it is always N-byte aligned
-struct AlignedBuffer<const N: usize>(NonNull<[u8; N]>);
-
-impl<const N: usize> AlignedBuffer<N> {
-    fn layout() -> Layout {
-        Layout::from_size_align(N, N).unwrap()
-    }
-
-    fn new() -> Self {
-        let p = unsafe { alloc_zeroed(Self::layout()) } as *mut [u8; N];
-        Self(NonNull::new(p).unwrap())
-    }
-
-    fn buf(&mut self) -> &mut [u8; N] {
-        unsafe { self.0.as_mut() }
-    }
-}
-
-impl<const N: usize> Drop for AlignedBuffer<N> {
-    fn drop(&mut self) {
-        unsafe { dealloc(self.0.as_ptr() as *mut u8, Self::layout()) }
-    }
-}
+use std::mem::MaybeUninit;
 
 // Used to benchmark the throughput of getrandom in an optimal scenario.
 // The buffer is hot, and does not require initialization.
 #[inline(always)]
-fn bench<const N: usize>(b: &mut test::Bencher) {
-    let mut ab = AlignedBuffer::<N>::new();
-    let buf = ab.buf();
+fn bench_getrandom<const N: usize>(b: &mut test::Bencher) {
+    b.bytes = N as u64;
     b.iter(|| {
+        let mut buf = [0u8; N];
         getrandom::getrandom(&mut buf[..]).unwrap();
         test::black_box(&buf);
     });
-    b.bytes = N as u64;
 }
 
 // Used to benchmark the throughput of getrandom is a slightly less optimal
 // scenario. The buffer is still hot, but requires initialization.
 #[inline(always)]
-fn bench_with_init<const N: usize>(b: &mut test::Bencher) {
-    let mut ab = AlignedBuffer::<N>::new();
-    let buf = ab.buf();
+fn bench_getrandom_uninit<const N: usize>(b: &mut test::Bencher) {
+    b.bytes = N as u64;
     b.iter(|| {
-        for byte in buf.iter_mut() {
-            *byte = 0;
-        }
-        getrandom::getrandom(&mut buf[..]).unwrap();
+        // TODO: When the feature `maybe_uninit_as_bytes` is available, use:
+        // ```
+        // let mut buf: MaybeUninit<[u8; N]> = MaybeUninit::uninit();
+        // getrandom::getrandom_uninit(buf.as_bytes_mut()).unwrap();
+        // test::black_box(unsafe { buf.assume_init() })
+        // ```
+        // since that is the shape we expect most callers to have.
+        let mut buf = [MaybeUninit::new(0u8); N];
+        let buf = getrandom::getrandom_uninit(&mut buf[..]).unwrap();
```

Comment:
The benchmarks work only on Nightly, so I think we can use [`MaybeUninit::uninit_array()`](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.uninit_array). IIUC this code performs zero initialization of the buffer, so no wonders you see no difference in the measured values.

Also, maybe it's better to consume the result directly by black box without unwrapping it?

Reply:
Doh! Thanks for catching that. Since I think `uninit_array()` is unlikely to ever get stabilized based on its documentation, I did something slightly different but equivalent.

Reply:
> Also, maybe it's better to consume the result directly by black box without unwrapping it?

For good performance we want the compiler to never actually construct the `Result`, so we don't want to consume the `Result`.

I did change the `black_box` calls from `black_box(&buf)` to `black_box(buf)` because I don't think `black_box(&buf)` is expected to really do much to mitigate unwanted optimizations. However, perhaps this will just create additional noise. Not sure what the right solution is.

Reply:
I think it's fine to do:
```rust
let mut uninit = [MaybeUninit::uninit(); N];
let buf = getrandom::getrandom_uninit(&mut uninit).unwrap();
test::black_box(buf);
```
The `getrandom_uninit` function gets inlined anyway, and `buf` is a reference, so I don't think the difference between `black_box(buf)` and `black_box(buf)` matters.

Reply:
This was already merged, but for posterity: we need to make sure that both variants of the benchmark function end up constructing a random `[u8; N]`, which is why I had to add the step to `assume_init()`.

Further I would love to know in more detail whether `black_box(&some_array)` can be expected to work the same as `black_box(some_array)`. I could see arguments for and against.

Thanks for merging!

Reply:
This is a good point. I can confirm that the value you pass to `black_box` does (in practice) change the emitted code. So I think we just need to make sure that if we are _comparing_ two different implementations, they should be passing the same type to `black_box`.